src/accel/detail/RootOffloadWriter.cc

```diff
@@ -0,0 +1,79 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file accel/detail/RootOffloadWriter.cc
+//---------------------------------------------------------------------------//
+#include "RootOffloadWriter.hh"
+
+#include <TFile.h>
+#include <TTree.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+#include "orange/Types.hh"
+#include "celeritas/phys/ParticleParams.hh"  // IWYU pragma: keep
+
+namespace celeritas
+{
+namespace detail
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Convert celeritas::Real3 to std::array<double, 3>.
+ */
+std::array<double, 3> real3_to_array(Real3 const& src)
+{
+    std::array<double, 3> dst;
+    std::memcpy(&dst, &src, sizeof(src));
+    return dst;
+}
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with ROOT output filename.
+ */
+RootOffloadWriter::RootOffloadWriter(std::string const& root_output_name,
+                                     SPConstParticles params)
+    : tfile_mgr_(root_output_name.c_str()), params_(std::move(params))
+{
+    ttree_ = tfile_mgr_.make_tree(this->tree_name(), this->tree_name());
+    ttree_->Branch("event_id", &primary_.event_id);
+    ttree_->Branch("track_id", &primary_.track_id);
+    ttree_->Branch("particle", &primary_.particle);
+    ttree_->Branch("energy", &primary_.energy);
+    ttree_->Branch("time", &primary_.time);
+    ttree_->Branch("pos", &primary_.pos);
+    ttree_->Branch("dir", &primary_.dir);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Export primaries to ROOT.
+ */
+void RootOffloadWriter::operator()(Primaries const& primaries)
+{
+    CELER_EXPECT(!primaries.empty());
+
+    std::scoped_lock{write_mutex_};
```

Comment:
You should remove this. The `OffloadWriter` is the only thing that needs to be thread safe. If we need thread safety outside what ROOT provides we can use the existing OffloadWriter and make its `EventWriter` a generic `std::function<void(vector<Primary>)>`. Also note that the way the reader is structured, you **MUST** lock out the file when writing an event, otherwise primaries from different events could be interleaved.

Reply:
Just to be sure I got your point: so you're going to lock the `operator()` operation  itself, whether that's a HepMC3 or ROOT writer?

Reply:
Right now the `OffloadWriter` in accel has a single purpose: to wrap an `EventWriter` with a thread safe lock. I think we'll use that same mechanism to wrap the RootWriter, since for the reading to work, we have to write all tracks consecutively for a single event.